src/mdesktopentry.cpp

```diff
@@ -244,14 +249,37 @@ MDesktopEntryPrivate::~MDesktopEntryPrivate()
 {
 }
 
-QTranslator *MDesktopEntryPrivate::loadTranslator() const
+std::shared_ptr<QTranslator> MDesktopEntryPrivate::loadTranslator() const
 {
-    QTranslator *translator = new QTranslator;
-    QString catalog = keyFile.stringValue(DesktopEntrySection, TranslationCatalogKey);
+    if (translatorUnavailable)
+        return std::shared_ptr<QTranslator>();
+
+    std::shared_ptr<QTranslator> translator = cachedTranslator.lock();
+    if (translator)
+        return translator;
+
+    translator = std::make_shared<QTranslator>();
+
+    QString catalog;
+    if (keyFile.contains(DesktopEntrySection, TranslationCatalogKey))
+        catalog = keyFile.stringValue(DesktopEntrySection, TranslationCatalogKey);
+    else if (keyFile.contains(DesktopEntrySection, LegacyTranslationCatalogKey))
+        catalog = keyFile.stringValue(DesktopEntrySection, LegacyTranslationCatalogKey);
     if (catalog.isNull() || !translator->load(QLocale(), catalog, "-", "/usr/share/translations")) {
         qDebug() << "Unable to load catalog" << catalog;
-        delete translator;
-        translator = 0;
+        translator.reset();
+        translatorUnavailable = true;
+    }
+
+    if (translator) {
+        cachedTranslator = translator;
+        translatorCacheTimer.reset(new QTimer);
+        translatorCacheTimer->setSingleShot(true);
+        translatorCacheTimer->setInterval(TranslatorLifetime);
+        QObject::connect(translatorCacheTimer.data(), &QTimer::timeout, [this, translator] {
+            translatorCacheTimer.reset();
+            // end of this scope: translator shared pointer is dropped -> cached QTranslator is deleted
```

Comment:
A bit mixed feelings with the unloading. In a way good if the QTranslator is not kept in memory forever (even though it shouldn't be that expensive), but the std:weak/shared pointers and the timer with scope end feel somewhat complicated.

For alternative there's been in many places QBasicTimer + timerEvent() which could provide slightly simpler and qt-like cleanup.

But guess this works too.

Reply:
Maybe this is slightly too fancy. :hand_over_mouth: I'll check that other approach still.

Reply:
Hmm.. How would you use `timerEvent()` when `MDesktopEntry` is not a `QObject` derived class? :thinking:

Reply:
I think glib provides the same functionality as `QBasicTimer`, using `g_timeout_add()` and cancel it based on the id with `g_source_remove()` but that depends on use of glib main loop (in Qt). Would that be okay to use here?

Reply:
Since that would need some extra glue, would it provide much more than just creating a QTimer here?

But for simpler solution, why not to just use plain(er) pointers and explicitly delete/reset the QTranslator instance on QTimer timeout?

Reply:
> Since that would need some extra glue, would it provide much more than just creating a QTimer here?

Well yeah, that would take twice the lines.

> But for simpler solution, why not to just use plain(er) pointers and explicitly delete/reset the QTranslator instance on QTimer timeout?

I guess I just wanted to avoid any possibility of races anywhere but if things are used the way one would think, then there is none. Replacing smart pointers with regular pointers and `QScopedPointer` results in the same number of rows minus `<memory>` include. :smile: